src/rulesets/Statistics/statistics.jl

```diff
@@ -6,19 +6,27 @@ _denom(x, dims::Colon) = length(x)
 _denom(x, dims::Integer) = size(x, dims)
 _denom(x, dims) = mapreduce(i->size(x, i), Base.mul_prod, unique(dims), init=1)
 
-# TODO: We have `mean(f, x; dims)` as of 1.3.0-DEV.36
-# https://github.com/JuliaDiff/ChainRules.jl/issues/85
-function rrule(::typeof(mean), x::AbstractArray{<:Real}; dims=:)
-    y_sum, sum_pullback = rrule(sum, x; dims=dims)
+function rrule(::typeof(mean), x::AbstractArray{<:Union{Real,Complex,AbstractArray}}; dims=:)
+    y_sum, sum_pullback = rrule(sum, x; dims)
     n = _denom(x, dims)
     function mean_pullback(ȳ)
-        _, ∂sum_x = sum_pullback(ȳ)
-        ∂x = unthunk(∂sum_x) / n
+        _, ∂x = sum_pullback(unthunk(ȳ) / n)
         return (NoTangent(), ∂x)
     end
     return y_sum / n, mean_pullback
 end
 
+function rrule(
+    config::RuleConfig{>:HasReverseMode},
+    ::typeof(mean),
+    f::F,
+    x::AbstractArray{T};
+    dims = :,
+) where {F, T<:Union{Real,Complex,AbstractArray}}
+    λ = one(real(eltype(T))) / _denom(x, dims)
+    # λ = 1 // _denom(x, dims)
+    y, back = rrule(config, sum, f, x; dims, _lambda = λ)
```

Comment:
Would it be possible to avoid passing in lambda if we do a similar thing than above? i.e. essentially saying that `mean(f, xs) = 1/length(xs) * sum(f, xs)`?

```suggestion
    N = length(xs)
    y_sum, pb_sum = rrule(config, sum, f, x; dims)
    function mean_pullback(ȳ)
        return pb_sum(N * ȳ)
    end
    return y_sum / N, mean_pullback
```
that means we can avoid passing the lambda and modifying the `rrule` for `sum`

Am I missing something here?

Reply:
It's been a while since I wrote this, so I forgot all the details. But I thought this had difficulty getting the gradient with `f` right, or something? Could be wrong though.

The other implementation I had stashed was this, but (IIRC) it had inference problems & ended up much slower:
```
function rrule(
    config::RuleConfig{>:HasReverseMode},
    ::typeof(mean),
    f::F,
    x::AbstractArray{T};
    dims = :,
) where {F, T<:Number}
    λ = float(one(T)) / _denom(x, dims)
    y, back = rrule(config, sum, Base.Fix1(*, λ) ∘ f, x; dims)
    function mean_pullback_f(∂y)
        _, ∂λf, ∂x = back(∂y)
        return (NoTangent(), ∂λf.inner, ∂x)
    end
    return y, mean_pullback_f
end

# Enable that to hit the efficient path despite more complicated `f`:
function ChainRulesCore.derivatives_given_output(
    ::Nothing,  # matches what rrule(config, sum, ...) uses
    λf::ComposedFunction{<:Base.Fix1{typeof(*), <:Number}, F},
    x::Number
) where {F}
    f = λf.inner
    λ = λf.outer.x
    g = only(only(derivatives_given_output(nothing, f, x)))
    return ((g * λ,),)
end
```

Reply:
That is much simpler, and seems to work locally. Not sure why I rejected it, must have had something else wrong.